File: src/main/kotlin/xyz/mahmoudahmed/dsl/TypedBranchActionBuilder.java
package xyz.mahmoudahmed.dsl;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;

public class TypedBranchActionBuilder<S, T> {
    private final TypedBranchBuilder<S, T> parent;
    private final Predicate<Object> condition;
    private final List<MappingRule> actions = new ArrayList<>();

    TypedBranchActionBuilder(TypedBranchBuilder<S, T> parent, Predicate<Object> condition) {
        this.parent = parent;
        this.condition = condition;
    }

    public <V> TypedBranchMappingBuilder<S, T, V> map(Function<S, V> sourceAccessor) {
        return new TypedBranchMappingBuilder<>(this, sourceAccessor);
    }

    public TypedBranchBuilder<S, T> endBranch() {
        parent.addBranch(new ConditionalBranch(condition, actions));
        return parent;
    }

    void addAction(MappingRule rule) {
        actions.add(rule);
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/typed/TypedBranchActionBuilder.kt
package xyz.mahmoudahmed.dsl.typed

import xyz.mahmoudahmed.dsl.conditional.ConditionalBranch
import xyz.mahmoudahmed.dsl.core.MappingRule

class TypedBranchActionBuilder<S, T>(
    private val parent: TypedBranchBuilder<S, T>,
    private val condition: (Any) -> Boolean
) {
    private val actions = mutableListOf<MappingRule>()

    fun <V> map(sourceAccessor: (S) -> V): TypedBranchMappingBuilder<S, T, V> {
        return TypedBranchMappingBuilder(this, sourceAccessor)
    }

    fun endBranch(): TypedBranchBuilder<S, T> {
        parent.addBranch(ConditionalBranch(condition, actions))
        return parent
    }

    internal fun addAction(rule: MappingRule) {
        actions.add(rule)
    }
}